Consolidate CLI tests and integration tests helpers

honzajavorek opened the pull request:
#### :rocket: Why this change?

Dredd's integration and regression tests often run Dredd in subprocess, lookup processes, kill processes, etc. However, at the moment almost every test file takes its own approach on how to do that. In order to support running the test suite on Windows, it's necessary to unify all approaches and have a common set of helpers. Unified helpers are then implemented with all OS-specific stuff in mind and ensure consistency and readability across the whole integration test suite.

Changes made in this PR:

- Existing helpers for CLI tests were moved from `test/integration/cli/helpers.coffee` to `test/integration/helpers.coffee`, so they could be easily accessed by any integration tests.
- Regression tests were moved from `test/regressions` to `test/integration/regressions`, so they could easily access the common helpers.
- All helpers were rewritten to be platform-independent (replacing `bash` process lookup and process killing with a cross-platform library, using `cross-spawn` everywhere, etc.).
- All helpers made as versatile as possible so all integration tests could utilize them.
- All integration and regression tests were refactored to use the new set of helpers. This often resulted in much shorter, cleaner, DRYier code.

#### :memo: Related issues and Pull Requests

- https://github.com/apiaryio/dredd/issues/204

#### :white_check_mark: What didn't I forget?

- [ ] To write docs - N/A
- [x] To write tests
- [x] To put [Conventional Changelog](https://dredd.readthedocs.io/en/latest/contributing/#sem-rel) prefixes in front of all my commits and run `npm run lint`

honzajavorek commented:
@abtris Adding a dependency: [ps-node](https://www.npmjs.com/package/ps-node)

honzajavorek commented:
@miiila This one is a bit hairy, but very necessary cleanup of integration tests. I offer pairing for the review 👨‍💻 🗯 👨‍💻 💬

Also cookies 🍪, beers 🍻, or other goods 🎁 can be in the game if we're able to get this to `master` during Monday 😼 🙏 🗓

miiila commented on test/integration/cli/api-description-cli-test.coffee:
```diff
@@ -33,221 +36,220 @@ describe 'CLI - API Description Document', ->
     app.post '/machine-types', (req, res) ->
       res.send [{name: 'bulldozer'}]
```
Omitted `res.send`; consistency check.

honzajavorek replied:
Addressed in cf601cd.

miiila commented on test/integration/cli/cli-test.coffee:
```diff
@@ -1,720 +1,618 @@
+          "http://127.0.0.1:#{DEFAULT_SERVER_PORT}"
+          '--path=./test/fixtures/multifile/*.apib'
+          '--only=Message API > /message > GET'
+          '--no-color'
```
Could you please verify whether `--no-color` is really needed?

honzajavorek replied:
Addressed in 1de2b69

miiila commented on test/integration/cli/cli-test.coffee:
```diff
@@ -1,720 +1,618 @@
+          './test/fixtures/single-get.apib'
+          "http://127.0.0.1:#{DEFAULT_SERVER_PORT}"
+          '--color'
+          'false'
```
not following GNU convention

honzajavorek replied:
Addressed in 47ec552

miiila commented on test/integration/cli/cli-test.coffee:
```diff
@@ -758,186 +651,190 @@ describe 'CLI', () ->
+    it 'should hit the resource', ->
+      assert.deepEqual runtimeInfo.server.requestCounts, {'/machines': 1}
+
+    it 'exit status should be 0', ->
```
I'd say exit status should not be `0`, when there is an error message expected.

honzajavorek replied:
Addressed in bd27e5f

miiila commented on test/integration/sanitation-test.coffee:
```diff
@@ -48,29 +48,30 @@ describe('Sanitation of Reported Data', ->
-        [err, results] = args
-        done(err)
+      runDreddWithServer(dredd, app, (err, runtimeInfo) ->
+        return done(err) if err
```
You are usually skipping these returns and sticking to `done(err)`; I'd keep it for the sake of the consistency.

honzajavorek replied:
It's not always possible. There are two different situations:

1. The callback receives `(err, object)` and I need the `object`:

```coffee
object = undefined
fn((err, obj) ->
  object = obj
  done(err)
)
```

2. The callback receives `(err, object)` and I need a property of the `object`:

```coffee
property = undefined
fn((err, obj) ->
  return done(err) if err
  property = obj.property
  done()
)
```

If I skipped the `return done(err) if err` in the second case, the line `property = obj.property` would blow up 💥 (accessing a `property` on `null` or `undefined`).

honzajavorek commented on test/integration/cli/api-description-cli-test.coffee:
```diff
@@ -33,221 +36,220 @@ describe 'CLI - API Description Document', ->
     describe 'When loaded by glob pattern', ->
-      dreddCommand = undefined
-      args = ['./test/fixtures/single-g*t.apib', "http://127.0.0.1:#{PORT}"]
+      dreddCommandInfo = undefined
```
Applies also for

- `api-description-cli-test.coffee`
- `api-blueprint-cli-test.coffee`
- `swagger-cli-test.coffee`

honzajavorek replied:
Addressed in 2dfdca3

honzajavorek commented:
@miiila Thanks for the review! I'm waiting for @abtris for his review of dependencies, so I guess I'm just going to add commits addressing your comments.

@abtris Any ETA on getting review from you?

honzajavorek commented:
@miiila Added several commits. OK to merge?

miiila commented:
Awesome! 🉑
